File: Coding_Problems/13_findFirstLastPos.js
function searchRange(nums, target) {
  return [findFirst(nums, target), findLast(nums, target)];
}

function findFirst(nums, target) {
  let left = 0,
    right = nums.length - 1;
  let result = -1;

  while (left <= right) {
    let mid = left + Math.floor((right - left) / 2);

    if (nums[mid] === target) {
      result = mid; // found it, but keep searching LEFT
      right = mid - 1; // ← this is what makes it "find first"
    } else if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return result;
}

function findLast(nums, target) {
  let left = 0,
    right = nums.length - 1;
  let result = -1;

  while (left <= right) {
    let mid = left + Math.floor((right - left) / 2);

    if (nums[mid] === target) {
      result = mid; // found it, but keep searching RIGHT
      left = mid + 1; // ← this is what makes it "find last"
    } else if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return result;
}

console.log(searchRange([5, 7, 7, 8, 8, 10], 8)); // [3, 4]
console.log(searchRange([5, 7, 7, 8, 8, 10], 6)); // [-1, -1]

File: Coding_Problems/CPP/13_findFirstLastPos.cpp
#include <iostream>
#include <vector>

using namespace std;

int findFirst(vector<int> &nums, int target)
{
    int left = 0;
    int right = nums.size() - 1;
    int result = -1;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;

        if (nums[mid] == target)
        {
            result = mid;
            right = mid - 1;
        }
        else if (nums[mid] < target)
        {
            left = mid + 1;
        }
        else
        {
            right = mid - 1;
        }
    }

    return result;
}

int findLast(vector<int> &nums, int target)
{
    int left = 0;
    int right = nums.size() - 1;
    int result = -1;

    while (left <= right)
    {
        int mid = left + (right - left) / 2;

        if (nums[mid] == target)
        {
            result = mid;
            left = mid + 1;
        }
        else if (nums[mid] < target)
        {
            left = mid + 1;
        }
        else
        {
            right = mid - 1;
        }
    }

    return result;
}

vector<int> searchRange(vector<int> &nums, int target)
{
    return {findFirst(nums, target), findLast(nums, target)};
}

int main()
{
    vector<int> nums = {8, 7, 7, 8, 10};

    vector<int> result = searchRange(nums, 8);

    cout
        << result[0] << ", " << result[1];

    return 0;
}
